### Script

- runthreepclr_genomicdata_split.sh

-- script to run 3P-CLR source code extended by JS, available at https://github.com/joshuamschmidt/3P-CLR

### Data

- 3pclr.scores.genomewide.chr.snpposition.1.txt

-- output of 3P-CLR on chimpanzee data - filtered to 3 columns

-- contains 3 columns: chromosome position 3P-CLR score

-- where position = physical position of the central SNP of the window

-- note this file contains 3P-CLR scores genome-wide (for all windows for which the statistic was calculated)

-- file used to plot Figure S12: Empirical distribution of 3P-CLR scores
  
To plot in R: 

```  
fulldist<-fread("3pclr.scores.genomewide.chr.snpposition.1.txt")
ggplot(fulldist, aes(x=V3)) +  geom_histogram() +  labs(x="3P-CLR score", y = "Frequency") + theme_classic() + theme(text = element_text(size = 15))  + theme(axis.title = element_text(size = 15)) + geom_vline(xintercept = quantile(fulldist$V3, probs = c(0.995))) + geom_vline(xintercept = quantile(fulldist$V3, probs = c(0.999))) + geom_vline(xintercept = quantile(fulldist$V3, probs = c(0.9995)))
 ```

#head of file

#15	6.18136e+07	693.253

#12	1.10856e+08	679.884

- noheader_aligned_3pclr_out_genome_wide_*percentsubset.txt

--  3 files - of the highest tails of 3P-CLR scores: the 0.5% (n=4090), 0.1% (n=818) and 0.05% (n=409) tails of the empirical distribution.

-- these files were used in downstream analysis

-- & plotted in Figure S13: Distribution of 3P-CLR scores in the tails of the empirical distribution

-- no header in files, but columns correspond to: Chr\tPhyspos\tGenpos\tPhysStart\tPhysEnd\tGenStart\tGenEnd\tn.snps\t3PCLR.Anc\t3PCLR.Anc.S\t3PCLR.A\t3PCLR.A.S\t3PCLR.B\t3PCLR.B.S

quantiles generated by
```
fulldist[(fulldist$V3 > quantile(fulldist$V3, probs = c(0.995))),]
fulldist[(fulldist$V3 > quantile(fulldist$V3, probs = c(0.999))),]
fulldist[(fulldist$V3 > quantile(fulldist$V3, probs = c(0.9995))),]
```
